browser/ui/views/brave_ads/ad_notification_popup_handler.cc

```diff
@@ -16,12 +19,37 @@ AdNotificationPopupHandler::~AdNotificationPopupHandler() = default;
 // static
 void AdNotificationPopupHandler::Show(Profile* profile,
                                       const AdNotification& ad_notification) {
-  AdNotificationPopup::Show(profile, ad_notification);
+  DCHECK(profile);
+
+  const std::string& id = ad_notification.id();
+
+  AdNotificationPopup* popup =
+      new AdNotificationPopup(profile, ad_notification);
+  AdNotificationPopupCollection::Add(id, popup);
+
+  AdNotificationDelegate* delegate = ad_notification.delegate();
+  if (delegate) {
+    delegate->OnShow();
+  }
 }
 
 // static
 void AdNotificationPopupHandler::Close(const std::string& notification_id) {
-  AdNotificationPopup::Close(notification_id, /* by_user */ false);
+  DCHECK(!notification_id.empty());
+
+  AdNotificationPopup* popup =
+      AdNotificationPopupCollection::Get(notification_id);
+  if (!popup) {
+    return;
+  }
+
+  const AdNotification ad_notification = popup->GetAdNotification();
+  AdNotificationDelegate* delegate = ad_notification.delegate();
+  if (delegate) {
+    delegate->OnClose(/* by_user */ false);
+  }
+
```

Comment:
`Add to collection` is done in Handler but `Erase from collection` is done in Popup.
How about calling `AdNotificationPopupCollection::Erase()` here also?

Reply:
Ah.. you mentioned the reason about this.
https://github.com/brave/brave-core/pull/10170/files#r713127450

It seems it's called sequentially from here. curious when race condition could happen.

Reply:
Previously erase from collection was done on widget destroy.
The race was in double `AdNotificationPopup::FadeOut()` call. This is sequence of what happened:
1) User presses popup and `AdNotificationView::OnMouseReleased()` is called
2) AdNotificationPopup::OnClick() call
3) direct AdNotificationPopup::FadeOut() call
4) AdNotificationDelegate::OnClick() call which led to `AdNotificationPopup::Close()` call by following chain:
`AdEventClicked::FireEvent() -> AdsClientHelper::Get()->CloseNotification() -> AdNotificationPopup::Close() -> AdNotificationPopup::Close()`
5) `AdNotificationPopup::Close()` found `AdNotificationPopup*` in collection and called AdNotificationPopup::FadeOut()
6) Due to animation widget was destroyed after some time and cleaned the collection

To fix this I did following:
- Not directly call `AdNotificationPopup::FadeOut()` from `AdNotificationView::OnMouseReleased()`
- Clear the collection earlier

Reply:
When could fade-out be called more than once?

Reply:
IIUC, `FadeOut()` **_was_** called twice from  `AdNotificationPopup::OnClick()` and `AdNotificationPopup::Close()` when user clicked ads view.
and firstly mentioned`FadeOut()` is deleted in this PR.

I think it's not much important but what I wanted to say was calling `AdNotificationPopupCollection::Erase()` here instead of calling it from `AdNotificationPopup::ClosePopup()` because `AdNotificationPopupCollection::Add` is also called in this handler.

Reply:
I got your point, will do that. Thank you!